Detectors/TOF/reconstruction/src/Encoder.cxx

```diff
@@ -0,0 +1,325 @@
+    PackedHit[iframe][phit].chain = summary[istart].getElChainIndex();
+    PackedHit[iframe][phit].tdcID = summary[istart].getElTDCIndex();
+    PackedHit[iframe][phit].channel = summary[istart].getElChIndex();
+    PackedHit[iframe][phit].time = hittimeTDC & 0x1FFF;
+    PackedHit[iframe][phit].tot = summary[istart].getTOT() /*bin 48.8 ns*/; // to be checked
+    nPackedHits[iframe]++;
+
+    if (mVerbose) {
+      auto Chain = PackedHit[iframe][phit].chain;
+      auto TDCID = PackedHit[iframe][phit].tdcID;
+      auto Channel = PackedHit[iframe][phit].channel;
+      auto Time = PackedHit[iframe][phit].time;
+      auto TOT = PackedHit[iframe][phit].tot;
+
+      int digitInfo[4];
+      int ext_time = iframe << 13;
+      Decoder::FromRawHit2Digit(icrate, itrm, TDCID, Chain, Channel, mOrbitID, mBunchID, Time + ext_time, TOT, digitInfo);
+
+      printf("orbit = %d\n", mOrbitID);
+      printf("DigitOriginal: channel = %d -- TDC = %d -- BC = %d -- TOT = %d \n", summary[istart].getChannel(), summary[istart].getTDC(), summary[istart].getBC(), summary[istart].getTOT());
+      printf("DigitAfter   :  channel = %d -- TDC = %d -- BC = %d -- TOT = %d \n", digitInfo[0], digitInfo[1], digitInfo[3], digitInfo[2]);
+    }
+
+    if (iframe < firstFilledFrame)
+      firstFilledFrame = iframe;
+    if (iframe > lastFilledFrame)
+      lastFilledFrame = iframe;
+
+    istart++;
+    if (istart < summary.size()) {
+      whatTRM = summary[istart].getElTRMIndex();
+      whatCrate = summary[istart].getElCrateIndex();
+    } else {
+      whatTRM = -1;
+      whatCrate = -1;
+    }
+  }
+
+  /** loop over frames **/
+  for (int iframe = firstFilledFrame; iframe < lastFilledFrame + 1; iframe++) {
+
+    /** check if frame is empty **/
+    if (nPackedHits[iframe] == 0)
+      continue;
+
+    // frame header
+    mUnion->frameHeader = {0x0};
+    mUnion->frameHeader.mustBeZero = 0;
+    mUnion->frameHeader.trmID = itrm;
+    mUnion->frameHeader.frameID = iframe;
+    mUnion->frameHeader.numberOfHits = nPackedHits[iframe];
+#ifdef VERBOSE
+    if (mVerbose) {
+      auto NumberOfHits = mUnion->frameHeader.numberOfHits;
+      auto FrameID = mUnion->frameHeader.frameID;
+      auto TRMID = mUnion->frameHeader.trmID;
+      //        std::cout << boost::format("%08x") % mUnion->data
+      //                  << " "
+      //                  << boost::format("Frame header (TRMID=%d, FrameID=%d, NumberOfHits=%d)") % TRMID % FrameID % NumberOfHits
+      //                  << std::endl;
+    }
+#endif
+    mUnion++;
+    mIntegratedBytes += 4;
+
+    // packed hits
+    for (int ihit = 0; ihit < nPackedHits[iframe]; ++ihit) {
+      mUnion->packedHit = PackedHit[iframe][ihit];
+      mUnion++;
+      mIntegratedBytes += 4;
+    }
+
+    nPackedHits[iframe] = 0;
+  }
+
+  // if current crate is over
+  if (whatCrate != icrate)
+    return -1;
+  // otherwise point to the next trm of the same crate
+  if (istart < summary.size())
```

Comment:
this should produce warning about comparison of signed with ulong, cast the size() to int.

Reply:
done